File: Examples/ObjectiveCExampleApp/ObjectiveCExampleApp/Controllers/PBBA/PBBAViewController.m
//
//  PBBAViewController.m
//  ObjectiveCExampleApp
//
//  Copyright (c) 2020 Alternative Payments Ltd
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

#import "PBBAViewController.h"
#import "UIViewController+Additions.h"

@import JudoKit_iOS;

@interface PBBAViewController () <JPPBBAButtonDelegate>
@property (nonatomic, strong) IBOutlet UILabel *orderIdSuffixLabel;
@property (nonatomic, strong) IBOutlet UILabel *orderIdValueLabel;
@property (weak, nonatomic) IBOutlet UIButton *checkStatusButton;
@property (strong, nonatomic) IBOutlet UIView *buttonPlaceholder;
@end

@implementation PBBAViewController

#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    if (JudoKit.isBankingAppAvailable) {
        [self createButtonProgrammatically];
    } else {
        self.orderIdSuffixLabel.text = @"No PBBA Bank App Found.";
        self.orderIdSuffixLabel.hidden = NO;
    }
}

- (IBAction)didTapCheckStatus:(id)sender {
    __weak typeof(self) weakSelf = self;
    JPApiService *apiService = [[JPApiService alloc] initWithAuthorization:self.judoKit.authorization
                                                               isSandboxed:self.judoKit.isSandboxed];
    [apiService invokeOrderStatusWithOrderId:self.orderIdValueLabel.text
                               andCompletion:^(JPResponse *response, JPError *error) {
                                   [weakSelf handleResponse:response error:error];
                               }];
}

#pragma mark - JPPBBAButtonDelegate

/**
 * Catch delegate call from JPPBBAButton
 *
 * @param sender - JPPBBAButton object, which triggers the click.
 */
- (void)pbbaButtonDidPress:(JPPBBAButton *)sender {
    __weak typeof(self) weakSelf = self;
    [weakSelf.judoKit invokePBBAWithConfiguration:weakSelf.configuration
                                       completion:^(JPResponse *response, JPError *error) {
                                           [weakSelf handleResponse:response error:error];
                                       }];
}

/**
 *  Method used for creating and adding JPPBBAButton Programmatically to view. Set up delegates
 */
- (void)createButtonProgrammatically {
    JPPBBAButton *pbbaFromCode = [[JPPBBAButton alloc] initWithFrame:self.buttonPlaceholder.bounds];
    pbbaFromCode.delegate = self;
    [self.buttonPlaceholder addSubview:pbbaFromCode];
    self.checkStatusButton.layer.cornerRadius = 4.0f;
}

#pragma mark - Handle Response

- (void)handleResponse:(JPResponse *)response error:(NSError *)error {
    if (error) {
        [self _jp_displayAlertWithError:error];
        return;
    }

    if (!response) {
        [self _jp_displayAlertWithError:[JPError _jp_requestFailedError]];
        return;
    }

    if (!response.orderDetails.orderStatus) {

        self.orderIdValueLabel.text = response.orderDetails.orderId;
        [self.orderIdValueLabel setHidden:!response.orderDetails.orderId];
        [self.orderIdSuffixLabel setHidden:!response.orderDetails.orderId];
        [self.checkStatusButton setHidden:!response.orderDetails.orderId];
        return;
    }

    self.configuration.pbbaConfiguration.deeplinkURL = nil;
    [self presentResultViewControllerWithResponse:response];
}

@end

File: Examples/SwiftExampleApp/SwiftExampleApp/Modules/PBBA/PBBAViewController.swift
//
//  PBBAViewController.swift
//  SwiftExampleApp
//
//  Copyright (c) 2020 Alternative Payments Ltd
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import UIKit
import JudoKit_iOS

class PBBAViewController: UIViewController, PBBAInteractorOutput {

    // MARK: - Constants

    private let kButtonHeight: CGFloat = 50.0
    private let kButtonWidth: CGFloat = 310.0
    private let kContentBottomOffset: CGFloat = 100.0

    // MARK: - Variables

    var interactor: PBBAInteractorInput!
    var coordinator: Coordinator?

    // MARK: - View lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        setupLayout()
        setupConstraints()
        setupTargets()
        interactor.viewDidLoad()
    }

    // MARK: - Protocol methods

    func navigateToResultsModule(with result: Result) {
        coordinator?.pushTo(.results(result))
    }

    func shouldDisplayTransactionStatusElements(_ shouldDisplay: Bool) {
        contentStackView.isHidden = !shouldDisplay
    }

    // MARK: - User actions

    @objc private func didTapGetTransactionButton() {
        interactor.checkTransactionStatus()
    }

    // MARK: - Layout setup

    private func setupLayout() {
        pbbaButton.delegate = self
        view.backgroundColor = .systemBackground
        view.addSubview(pbbaButton)

        contentStackView.addArrangedSubview(orderIDLabel)
        contentStackView.addArrangedSubview(getTransactionStatusButton)

        view.addSubview(contentStackView)
    }

    private func setupConstraints() {
        let constraints = [
            getTransactionStatusButton.widthAnchor.constraint(equalToConstant: kButtonWidth),
            getTransactionStatusButton.heightAnchor.constraint(equalToConstant: kButtonHeight),
            contentStackView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            contentStackView.bottomAnchor.constraint(equalTo: view.bottomAnchor,
                                                     constant: -kContentBottomOffset)
        ]
        NSLayoutConstraint.activate(constraints)
    }

    private func setupTargets() {
        getTransactionStatusButton.addTarget(self,
                                             action: #selector(didTapGetTransactionButton),
                                             for: .touchDown)
    }

    // MARK: - Lazy instantiations

    private lazy var pbbaButton: JPPBBAButton = {
        let screenSize = UIScreen.main.bounds.size
        let originX = screenSize.width / 2 - kButtonWidth / 2
        let originY = screenSize.height / 2 - kButtonHeight / 2

        let frame = CGRect(x: originX,
                           y: originY,
                           width: kButtonWidth,
                           height: kButtonHeight)

        return JPPBBAButton(frame: frame)
    }()

    private lazy var orderIDLabel: UILabel = {
        let label = UILabel()
        label.numberOfLines = 0
        label.font = UIFont.systemFont(ofSize: 16.0, weight: .semibold)
        label.textColor = .systemGreen
        return label
    }()

    private lazy var getTransactionStatusButton: UIButton = {
        let button = UIButton()
        button.backgroundColor = .systemOrange
        button.setTitle("Check transaction status", for: .normal)
        button.setTitleColor(.label, for: .normal)
        button.layer.cornerRadius = 5.0
        return button
    }()

    private lazy var contentStackView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.axis = .vertical
        stackView.spacing = 20.0
        return stackView
    }()
}

extension PBBAViewController: JPPBBAButtonDelegate {

    func pbbaButtonDidPress(_ sender: JPPBBAButton) {
        interactor.didTapPBBAButton()
    }
}
